### ISSUE TITLE
ALPN set APIs return 0 on success, 1 on failure

### ISSUE THREAD

**Issue by tmshort** [openssl/openssl#16474]
So, they behave more like exit() values, rather than true/false for success/failure.

e.g.
SSL_CTX_set_alpn_protos()
SSL_set_alpn_protos()

I don't know how many people actually use the return value, but that should be a consideration before changing the APIs.

Do we want to fix these before 3.0 goes out? I realize it might be a bit late. Alternatively, add new _ex APIs that return more sensibly?

Based on the documentation, a number of other functions also return 0 on success, but they might have multiple return value meanings:
```
DES_random_key.pod:return 0 on success or negative values on error.
OSSL_CMP_CTX_new.pod:else 0 on successful validation,
SSL_CTX_set_alpn_select_cb.pod:SSL_CTX_set_alpn_protos() and SSL_set_alpn_protos() return 0 on success, and
SSL_CTX_set_tlsext_use_srtp.pod:SSL_CTX_set_tlsext_use_srtp() and SSL_set_tlsext_use_srtp() return 0 on success
UI_create_method.pod:return 0 on success, -1 if the given B<method> is NULL.
UI_new.pod:the UI is destroyed.  UI_dup_user_data() returns 0 on success, or -1 on memory
UI_new.pod:(Interrupt, Cancel, ...), -1 on error and 0 on success.
UI_new.pod:UI_dup_user_data() returns 0 on success or -1 on error.
UI_new.pod:UI_get_result_length() returns a positive integer or 0 on success; otherwise it
UI_new.pod:UI_process() returns 0 on success or a negative value on error.
UI_STRING.pod:UI_set_result() returns 0 on success or when the UI string is of any
UI_UTIL_read_pw.pod:UI_UTIL_read_pw_string() and UI_UTIL_read_pw() return 0 on success or a negative
X509_get0_signature.pod:X509_REQ_set1_signature_algo() returns 0 on success; or 1 on an
```

**Comment by davidben** [comment]
Changing the return value for the existing probably does not meet OpenSSL 3.0's compatibility goals. It's also the sort of thing that would only be noticed at runtime, so it's rather rude. But introducing more idiomatic return value, under a different name, sounds reasonable to me.

We did something similar for `SSL_CTX_set_tlsext_use_srtp` and called the new one `SSL_CTX_set_srtp_profiles`. Seems a good opportunity to drop the silly `tlsext` and `set_use_srtp` was very odd for something which took more than a boolean.

**Comment by tmshort** [comment]
Yup, `SSL_CTX_set_tlsext_use_srtp` shows up in that list...

**Comment by davidben** [comment]
Yeah, that's why I mentioned it. It's a good list. :-)

Looking at the list again, I think this one is just a documentation mistake, thankfully:

> X509_get0_signature.pod:X509_REQ_set1_signature_algo() returns 0 on success; or 1 on an

`X509_REQ_set1_signature_algo` just calls `X509_ALGOR_copy`, which returns 1 on success and 0 on error.

**Comment by mattcaswell** [comment]
> ALPN set APIs return 0 on success, 1 on failure

Yes, I'm fairly sure there are more functions like this - which is irritating. Too late for 3.0 though IMO.

> Changing the return value for the existing probably does not meet OpenSSL 3.0's compatibility goals. It's also the sort of thing that would only be noticed at runtime, so it's rather rude. But introducing more idiomatic return value, under a different name, sounds reasonable to me.

+1

**Comment by tmshort** [comment]
I think we all agree that existing APIs really *can't* be changed, but that consistent return values are useful.

Thinking about this further, my gut is telling me that APIs need to be *even* more consistent. It seems there's a bunch of inconsistency in certain types of APIs, and return values.

e.g. 
How to get/set string values. (In general how strings are handled, e.g. move everything internally to `BUF_MEM` and use that with get1/get0 operations?)
Consistent use of `size_t`/`ssize_t`(vs `unsigned int` or `int` or even `long`)
Callbacks and other settings and their propagation (set on SSL_CTX, set on SSL, copied from SSL_CTX -> SSL, referenced from SSL -> SSL_CTX)

**Comment by nhorman** [comment]
To make concrete progress on this, it would be helpful to have an enumerated list of API's that are inconsistent, so that we could look at addressing them in the 4.0 time frame

**Comment by mattcaswell** [comment]
> To make concrete progress on this, it would be helpful to have an enumerated list of API's that are inconsistent, so that we could look at addressing them in the 4.0 time frame

@tmshort provided such a list in the original description of this issue - although each one needs to be validated to confirm they really are a problem.

**Comment by nhorman** [comment]
Ok, moving to consideration
